source/game/system/RaceManager.cc

```diff
@@ -150,24 +163,76 @@ void RaceManager::Player::calc() {
     if (m_checkpointFactor < 0.0f || m_checkpointId != checkpointId) {
         calcCheckpoint(checkpointId, distanceRatio);
     }
+
+    m_raceCompletion = static_cast<f32>(m_currentLap) +
+            (m_checkpointStartLapCompletion + m_checkpointFactor * distanceRatio);
+    m_raceCompletion = std::min(m_raceCompletion, m_currentLap + 0.99999f);
+}
+
+u16 RaceManager::Player::checkpointId() const {
+    return m_checkpointId;
+}
+
+f32 RaceManager::Player::raceCompletion() const {
+    return m_raceCompletion;
+}
+
+s8 RaceManager::Player::jugemId() const {
+    return m_jugemId;
 }
 
 const KPad *RaceManager::Player::inputs() const {
     return m_inputs;
 }
 
 /// @addr{0x80534DF8}
-MapdataCheckPoint *RaceManager::Player::calcCheckpoint(u16 checkpointId, f32 /*distanceRatio*/) {
+MapdataCheckPoint *RaceManager::Player::calcCheckpoint(u16 checkpointId, f32 distanceRatio) {
     auto *courseMap = CourseMap::Instance();
+
+    u16 oldCheckpointId = m_checkpointId;
     m_checkpointId = checkpointId;
+
+    f32 lapProportion = courseMap->checkPath()->lapProportion();
     MapdataCheckPath *checkPath = courseMap->checkPath()->findCheckpathForCheckpoint(checkpointId);
-    m_checkpointFactor = checkPath->oneOverCount() * courseMap->checkPath()->lapProportion();
+    m_checkpointFactor = checkPath->oneOverCount() * lapProportion;
+
+    m_checkpointStartLapCompletion = static_cast<f32>(checkPath->depth()) * lapProportion +
+            (m_checkpointFactor * static_cast<f32>(checkpointId - checkPath->start()));
+
+    f32 deltaLapCompletion =
+            m_lapCompletion - (m_checkpointStartLapCompletion + distanceRatio * m_checkpointFactor);
 
-    return courseMap->getCheckPoint(checkpointId);
+    MapdataCheckPoint *newCheckpoint = courseMap->getCheckPoint(checkpointId);
+    const MapdataCheckPoint *oldCheckpoint = courseMap->getCheckPoint(oldCheckpointId);
+
+    s8 newJugemIdx = newCheckpoint->jugemIndex();
+    if (newJugemIdx != -1) {
```

Comment:
```suggestion
    if (newJugemIdx >= 0) {
```
To address a question, it is very likely this is an `s8` due to the presence of the `extsb` instruction for comparison.